src/evaluators/BinaryExpression.js

```diff
@@ -190,17 +173,48 @@ export function computeBinary(
   }
 
   if (lval instanceof AbstractValue || rval instanceof AbstractValue) {
-    // generate error if binary operation might throw or have side effects
     if (realm.isInPureScope()) {
-      realm.evaluateWithPossibleThrowCompletion(
-        () => {
-          getPureBinaryOperationResultType(realm, op, lval, rval, lloc, rloc);
-          return realm.intrinsics.undefined;
-        },
-        TypesDomain.topVal,
-        ValuesDomain.topVal
-      );
+      // If we're in pure mode we can recover even if this operation might not be pure.
+      // To do that, we'll temporarily override the error handler.
+      const previousErrorHandler = realm.errorHandler;
+      let isPure = true;
+      realm.errorHandler = diagnostic => {
+        isPure = false;
+        return "Recover";
+      };
+      try {
+        const valueType = getPureBinaryOperationResultType(realm, op, lval, rval, lloc, rloc);
+        if (!isPure) {
+          // If this ended up reporting an error, it might not be pure, so we'll leave it in
+          // as a temporal operation with a known return type.
+
+          // Some of these values may trigger side-effectful user code such as valueOf.
+          // To be safe, we have to Havoc them. However, "in" and "instanceof" cannot
```

Comment:
I find the term "safe" here a bit misleading. It seems that in "pure" mode, we simply assume that unknown functions such as abstract functions with explicit calls, or in this case, implicit calls to unknown valueOf and toString functions, will refrain from side effecting the global state in general, but are permitted to leak and/or side-effect their arguments and hence we havoc those. It probably bears repeating this definition of "pure" mode all over the code base so that it is hard to miss.

Reply:
I think the notion of the "pure" mode is causing unnecessary confusion since it really shouldn't really need to be a mode at all.

It should be safer to always use this logic, whether you're in a pure scope or not. It is just that if you're not, the global object also needs to get havoced as soon as any unknown side-effect happens.

The purpose of the pure mode is just to scope the consequence of the havoc, not to change anything about the havoc logic such in this place.

This branching only fills two purposes: 1) Possible performance optimization. 2) To allow the normal (non-pure) mode to assume that these are pure and recover when they might not be for deeper optimizations (avoiding havoc). Maybe this should go into an "unsafe mode"?